File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool luhn_check(long cc_number);
void cc_type(long cc_number);

int main(void)
{
    // Checks if a given credit card number is valid, and if so, what kind of
    // credit card it is.

    // Get the number.
    long cc_number = get_long("Number: ");

    // Perform a Luhn check.
    if (luhn_check(cc_number))
    {
        cc_type(cc_number);
    }
    // Failed the Luhn check.
    else
    {
        printf("INVALID\n");
    }
}

bool luhn_check(long cc_number)
{
    // Performs a Luhn check on a credit card number.
    // Initialize the check sum.
    int check_sum = 0;
    // Initialize the position, starting from the right.
    int position = 1;
    long divisor = 1;
    while (divisor < cc_number)
    {
        int digit = (cc_number / divisor) % 10;
        // If we are in an even position, we need to multiply by 2.
        if (position % 2 == 0)
        {
            digit *= 2;
        }
        if (digit < 10)
        {
            check_sum += digit;
        }
        else
        {
            check_sum += (digit % 10);
            // If the doubled number is two digits, the first digit must be 1.
            check_sum += 1;
        }
        position++;
        divisor *= 10;
    }
    return (check_sum % 10 == 0);
}

void cc_type(long cc_number)
{
    // Use log to determine the length of the number.
    int length = (int)(log10(cc_number) + 1);
    // Get the first two digits of the credit card number.
    int start = (int)(cc_number / pow(10, length - 2));
    // Amex is 15 digits long and starts with 34 or 37.
    if (length == 15 && (start == 34 || start == 37))
    {
        printf("AMEX\n");
    }
    // MasterCard is 16 digits long and starts 51-55.
    else if (length == 16 && (start >= 51 && start <= 55))
    {
        printf("MASTERCARD\n");
    }
    // Visa is either 13 or 16 digits long and starts with 4.
    else if ((length == 13 || length == 16) && (start >= 40 && start <= 49))
    {
        printf("VISA\n");
    }
    // Else, the card is invalid.
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
def luhn_check(cc_number):
    """Performs a Luhn check on a credit card number using the Luhn algorithm.

    The Luhn algorithm doubles the value of every second digit, starting from
    t, and replaces that digit with the sum of that value. Then, all digits
    are summed together. If the sum is equal to 0 mod 10, the number passes
    the Luhn check.

    Args:
        cc_number: A credit card number.

    Returns:
        True if the number passes. False if it fails or the number is invalid.

    """
    cc_number = str(cc_number)
    cc_len = len(cc_number)
    check_sum = 0
    for i in range(cc_len):
        # Get the position from the right
        position = cc_len - i
        try:
            digit = int(cc_number[i])
        except ValueError:
            return False
        if position % 2 == 0:
            digit *= 2
            if digit >= 10:
                # If the sum is 2 digits (10-18), the sum of the digits is the
                # sum minus 9.
                digit -= 9
        check_sum += digit

    return check_sum % 10 == 0


def cc_type(cc_number):
    """Prints the type of a credit card number.

    Will print one of "AMEX", "VISA", or "MASTERCARD". If the type is none of
    these, the function prints "INVALID".

    Args:
        cc_number: A credit card number.

    """
    # Perform a Luhn check to make sure the number is valid:
    if not luhn_check(cc_number):
        print("INVALID")
        return

    cc_number = str(cc_number)
    cc_len = len(cc_number)
    # The first two digits
    start = cc_number[0:2]

    # AMEX starts with 34 or 37
    if cc_len == 15 and start in ['34', '37']:
        print("AMEX")
    # Mastercard starts with 51-55
    elif cc_len == 16 and start in [str(x) for x in range(51, 56)]:
        print("MASTERCARD")
    # Visa starts with 40-49
    elif cc_len == 16 and start in [str(x) for x in range(40, 50)]:
        print("VISA")
    else:
        print("INVALID")


if __name__ == '__main__':
    # Get a number.
    while True:
        cc_number = input("Number: ")
        try:
            int(cc_number)
            assert int(cc_number) >= 0
            break
        except (ValueError, AssertionError):
            pass

    cc_type(cc_number)
